File: zantt-api-backend/src/main/java/com/zantt/services/WorkspaceService.java
package com.zantt.services;

import com.zantt.entities.WorkspaceEntity;
import com.zantt.exceptions.WellKnownApiException;
import com.zantt.repositories.ProjectRepository;
import com.zantt.repositories.TaskRepository;
import com.zantt.repositories.WorkspaceRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
public class WorkspaceService {
    @Autowired
    WorkspaceRepository workspaceRepository;
    @Autowired
    TaskRepository taskRepository;

    public List<WorkspaceEntity> getWorkspacesByTaskId(String taskId) {
        var task = taskRepository.findById(taskId);
        if (task.isEmpty()) {
            throw new WellKnownApiException("Can not found task", "TASK_NOT_FOUND");
        }

        return workspaceRepository.findByTaskId(taskId);
    }

    public WorkspaceEntity addWorkspace(String taskId, String title, String contents) {
        var task = taskRepository.findById(taskId);
        if (task.isEmpty()) {
            throw new WellKnownApiException("Can not found task", "TASK_NOT_FOUND");
        }

        var workspaceId = UUID.randomUUID().toString();

        var workspace = new WorkspaceEntity();
        workspace.setWorkspaceId(workspaceId);
        workspace.setProjectId(task.get().getProjectId());
        workspace.setTaskId(taskId);
        workspace.setTitle(title);
        workspace.setContents(contents);

        return workspaceRepository.save(workspace);
    }

    public void deleteWorkspace(String workspaceId) {
        workspaceRepository.deleteById(workspaceId);
    }
}

File: zantt-api-backend-net/Zantt/Services/WorkspaceService.cs
using Zantt.Entities;
using Zantt.Repositories;

namespace Zantt.Services;

public class WorkspaceService
{
    private readonly WorkspaceRepository workspaceRepository;
    private readonly ILogger<WorkspaceService> logger;

    public WorkspaceService(
        ILogger<WorkspaceService> logger,
        WorkspaceRepository workspaceRepository)
    {
        this.logger = logger;
        this.workspaceRepository = workspaceRepository;
    }

    public List<WorkspaceEntity> GetWorkspaces(string taskId)
    {
        return workspaceRepository.GetWorkspacesByTaskId(taskId);
    }

    public WorkspaceEntity? GetWorkspace(string workspaceId)
    {
        return workspaceRepository.GetWorkspaceByWorkspaceId(workspaceId);
    }

    public WorkspaceEntity? AddWorkspace(string projectId, string taskId, string title, string contents)
    {
        return workspaceRepository.AddWorkspace(projectId, taskId, title, contents);
    }

    public WorkspaceEntity? UpdateWorkspace(string workspaceId, string title, string contents)
    {
        return workspaceRepository.UpdateWorkspace(workspaceId, title, contents);
    }

    public void DeleteWorkspace(string workspaceId)
    {
        workspaceRepository.DeleteWorkspace(workspaceId);
    }
}
